src/engine/sparqlExpressions/ConvertToNumericExpression.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024
+
+#include "engine/sparqlExpressions/NaryExpressionImpl.h"
+
+namespace sparqlExpression {
+namespace detail::to_numeric {
+
+// class that converts an input `int64_t`, `double` or `std::string`
+// to a numeric value `int64_t` or `double`
+template <typename T>
+class ToNumericImpl {
+ private:
+  Id getFromString(std::string& input) const {
+    auto str = absl::StripAsciiWhitespace(input);
+    auto strEnd = str.data() + str.size();
+    auto strStart = str.data();
+    T resT{};
+    if constexpr (std::is_same_v<T, int64_t>) {
+      auto conv = std::from_chars(strStart, strEnd, resT);
+      if (conv.ec == std::error_code{} && conv.ptr == strEnd) {
+        return Id::makeFromInt(resT);
+      }
+    } else {
+      auto conv = absl::from_chars(strStart, strEnd, resT);
+      if (conv.ec == std::error_code{} && conv.ptr == strEnd) {
+        return Id::makeFromDouble(resT);
+      }
+    }
+    return Id::makeUndefined();
+  };
+
+  Id getFromInt(int64_t input) const {
+    T num = static_cast<T>(input);
+    if constexpr (std::is_same_v<T, int64_t>) {
+      return Id::makeFromInt(num);
+    } else {
+      return Id::makeFromDouble(num);
+    }
+  };
+
+  Id getFromDouble(double input) const {
+    T num = static_cast<T>(input);
+    if constexpr (std::is_same_v<T, int64_t>) {
+      return Id::makeFromInt(num);
+    } else {
+      return Id::makeFromDouble(num);
+    }
+  };
```

Comment:
Those two functions have exactly the same body.
Make this function a template to reduce the code duplication:
```
template <typename T> requires (std::floating_point<T> || std::integral<T>)
Id getFromNumber(T number) const {...}
```
The concept with the requires clause is technically not necessary, but makes the code easier to read about which types are acceptable here.